README
NAME
    DBD::Oracle - Oracle database driver for the DBI module

VERSION
    version 1.75_2

SYNOPSIS
      use DBI;

      $dbh = DBI->connect("dbi:Oracle:$dbname", $user, $passwd);

      $dbh = DBI->connect("dbi:Oracle:host=$host;sid=$sid", $user, $passwd);

      # See the DBI module documentation for full details

      # for some advanced uses you may need Oracle type values:
      use DBD::Oracle qw(:ora_types);

DESCRIPTION
    DBD::Oracle is a Perl module which works with the DBI module to provide
    access to Oracle databases.

    This documentation describes driver specific behaviour and restrictions.
    It is not supposed to be used as the only reference for the user. In any
    case consult the DBI documentation first!

CONSTANTS
    :ora_session_modes
        ORA_SYSDBA ORA_SYSOPER ORA_SYSASM ORA_SYSBACKUP ORA_SYSDG ORA_SYSKM

    :ora_types
          ORA_VARCHAR2 ORA_STRING ORA_NUMBER ORA_LONG ORA_ROWID ORA_DATE ORA_RAW
          ORA_LONGRAW ORA_CHAR ORA_CHARZ ORA_MLSLABEL ORA_XMLTYPE ORA_CLOB ORA_BLOB
          ORA_RSET ORA_VARCHAR2_TABLE ORA_NUMBER_TABLE SQLT_INT SQLT_FLT ORA_OCI
          SQLT_CHR SQLT_BIN

    SQLCS_IMPLICIT
    SQLCS_NCHAR
        SQLCS_IMPLICIT and SQLCS_NCHAR are *character set form* values. See
        notes about Unicode elsewhere in this document.

    SQLT_INT
    SQLT_FLT
        These types are used only internally, and may be specified as
        internal bind type for ORA_NUMBER_TABLE. See notes about
        ORA_NUMBER_TABLE elsewhere in this document

    ORA_OCI
        Oracle doesn't provide a formal API for determining the exact
        version number of the OCI client library used, so DBD::Oracle has to
        go digging (and sometimes has to more or less guess). The ORA_OCI
        constant holds the result of that process.

        In string context ORA_OCI returns the full "A.B.C.D" version string.

        In numeric context ORA_OCI returns the major.minor version number
        (8.1, 9.2, 10.0 etc). But note that version numbers are not actually
        floating point and so if Oracle ever makes a release that has a two
        digit minor version, such as 9.10 it will have a lower numeric value
        than the preceding 9.9 release. So use with care.

        The contents and format of ORA_OCI are subject to change (it may,
        for example, become a *version object* in later releases). I
        recommend that you avoid checking for exact values.

    :ora_fetch_orient
          OCI_FETCH_CURRENT OCI_FETCH_NEXT OCI_FETCH_FIRST OCI_FETCH_LAST
          OCI_FETCH_PRIOR OCI_FETCH_ABSOLUTE OCI_FETCH_RELATIVE

        These constants are used to set the orientation of a fetch on a
        scrollable cursor.

    :ora_exe_modes
          OCI_STMT_SCROLLABLE_READONLY

    :ora_fail_over
          OCI_FO_END OCI_FO_ABORT OCI_FO_REAUTH OCI_FO_BEGIN OCI_FO_ERROR
          OCI_FO_NONE OCI_FO_SESSION OCI_FO_SELECT OCI_FO_TXNAL OCI_FO_RETRY

DBI CLASS METHODS
  connect
    This method creates a database handle by connecting to a database, and
    is the DBI equivalent of the "new" method. To open a connection to an
    Oracle database you need to specify a database connection string (URL),
    username and password.

    The connection string is always of the form: "dbi:Oracle:<db
    identifier>" There are several ways to identify a database:

    1   If the database is local, specifying the SID or service name will be
        enough.

    2   If the database is defined in a TNSNAMES.ORA file, you can use the
        service name given in the file

    3   To connect without TNSNAMES.ORA file, you can use an EZCONNECT url,
        of the form: //host[:port][/service_name]

    If port name is not specified, 1521 is the default. If service name is
    not specified, the hostname will be used as a service name.

    The following examples show several ways a connection can be created:

      $dbh = DBI->connect('dbi:Oracle:DB','username','password');

      $dbh = DBI->connect('dbi:Oracle:DB','username/password','');

      $dbh = DBI->connect('dbi:Oracle:','username@DB','password');

      $dbh = DBI->connect('dbi:Oracle:host=foobar;sid=DB;port=1521', 'scott/tiger', '');

      $dbh = DBI->connect("dbi:Oracle://myhost:1522/ORCL",'username', 'password');

   OS authentication
    To connect to a local database with a user which has been set up to
    authenticate via the OS ("ALTER USER username IDENTIFIED EXTERNALLY"):

      $dbh = DBI->connect('dbi:Oracle:','/','');

    Note the lack of a connection name (use the ORACLE_SID environment
    variable). If an explicit SID is used you will probably get an ORA-01004
    error.

    That only works for local databases. (Authentication to remote Oracle
    databases using your Unix login name without a password is possible but
    it is not secure and not recommended so not documented here.

   Oracle Environment Variables
    To use DBD::ORACLE to connect to an Oracle database, ORACLE_HOME
    environment variable should be set correctly. In general, the value used
    should match the version of Oracle that was used to build DBD::Oracle.
    If using dynamic linking then ORACLE_HOME should match the version of
    Oracle that will be used to load in the Oracle client libraries (via
    LD_LIBRARY_PATH, ldconfig, or similar on Unix).

    Oracle can use two environment variables to specify default connections:
    ORACLE_SID and TWO_TASK.

    To use them, specify either a local SID or service name, or a service
    name that is specified in the TNSNAMES.ORA file.

    Note that if you have *both* local and remote databases, and you have
    ORACLE_SID *and* TWO_TASK set, and you don't specify a fully qualified
    connect string on the command line, TWO_TASK takes precedence over
    ORACLE_SID (i.e. you get connected to remote system).

    It is highly recommended not to rely on environment variables and to
    always explicitly specify the SID in the connection string. This can
    prevent serious mistakes such as dropping a schema in the wrong
    database, and generally makes debugging and troubleshooting easier.

    Also remember that depending on the operating system you are using the
    differing "ORACLE" environment variables may be case sensitive, so if
    you are not connecting as you should double check the case of both the
    variable and its value.

   Timezones
    If the query is run through SQL*Net (mostly queries that are executed on
    remote servers), Oracle will return the time zone based on the setting
    of the UNIX environment variable "TZ" for the user who started the
    listener.

    If the query is run locally, Oracle will return the time zone based on
    the "TZ" environment variable setting of the user running the query.

    With local queries, you can change the time zone for a particular user
    by simply changing the setting of "TZ". To check the current setting,
    issue the UNIX "date" command.

   Oracle DRCP
    DBD::Oracle supports DRCP (Database Resident Connection Pool) so if you
    have an 11.2 database and DRCP is enabled you can direct all of your
    connections to it by adding ':POOLED' to the SID or setting a connection
    attribute of ora_drcp, or set the SERVER=POOLED when using a TNSENTRY
    style connection or even by setting an environment variable ORA_DRCP.
    All of which are demonstrated below;

      $dbh = DBI->connect('dbi:Oracle:DB:POOLED','username','password')

      $dbh = DBI->connect('dbi:Oracle:','username@DB:POOLED','password')

      $dbh = DBI->connect('dbi:Oracle:DB','username','password',{ora_drcp=>1})

      $dbh = DBI->connect('dbi:Oracle:DB','username','password',{ora_drcp=>1,
                                                                 ora_drcp_class=>'my_app',
                                                                 ora_drcp_min  =>10})

      $dbh = DBI->connect('dbi:Oracle:host=foobar;sid=ORCL;port=1521;SERVER=POOLED', 'scott/tiger', '')

      $dbh = DBI->connect('dbi:Oracle:', q{scott/tiger@(DESCRIPTION=
      (ADDRESS=(PROTOCOL=TCP)(HOST= foobar)(PORT=1521))
      (CONNECT_DATA=(SID=ORCL)(SERVER=POOLED)))}, "")

      if the ORA_DRCP environment variable is set then just this

      $dbh = DBI->connect('dbi:Oracle:DB','username','password')

    You can find a white paper on setting up DRCP and its advantages at
    <http://www.oracle.com/technetwork/articles/oracledrcp11g-1-133381.pdf>.

    Please note that DRCP support in DBD::Oracle is relatively new so the
    mechanics or its implementation are subject to change.

   TAF (Transparent Application Failover)
    Transparent Application Failover (TAF) is the feature in OCI that allows
    for clients to automatically reconnect to an instance in the event of a
    failure of the instance. The reconnect happens automatically from within
    the OCI (Oracle Call Interface) library. DBD::Oracle now supports a
    callback function that will fire when a TAF event takes place. You may
    use the callback to inform the user a failover is taking place or to
    setup the session again once the failover has succeeded.

    You will have to set up TAF on your instance before you can use this
    callback. You can test your instance to see if you can use TAF callback
    with

      $dbh->ora_can_taf();

    If you try to set up a callback without it being enabled DBD::Oracle
    will croak.

    NOTE: Currently, you must enable TAF during DBI's connect. However once
    enabled you can change the TAF settings.

    It is outside the scope of this document to go through all of the
    possible TAF situations you might want to set up but here is a simple
    example:

    The TNS entry for the instance has had the following added to the
    CONNECT_DATA section

       (FAILOVER_MODE=
                   (TYPE=select)
                   (METHOD=basic)
                   (RETRIES=10)
                   (DELAY=10))

    You will also have to create your own perl function that will be called
    from the client. You can name it anything you want and it will always be
    passed two parameters, the failover event value and the failover type.
    You can also set a sleep value in case of failover error and the OCI
    client will sleep for the specified seconds before it attempts another
    event.

      use DBD::Oracle(qw(:ora_fail_over));
      #import the ora fail over constants

      #set up TAF on the connection
      # NOTE since DBD::Oracle uses call_pv you may need to pass a full
      # name space as the function e.g., 'main::handle_taf'
      # NOTE from 1.49_00 ora_taf_function can accept a code ref as well
      #      as a sub name as it now uses call_sv
      my $dbh = DBI->connect('dbi:Oracle:XE', 'hr', 'hr',
                             {ora_taf_function => 'main::handle_taf'});

      #create the perl TAF event function

      sub handle_taf {
        # NOTE from 1.49_00 the $dbh handle was passed to your callback
        my ($fo_event,$fo_type, $dbh) = @_;
        if ($fo_event == OCI_FO_BEGIN){

          print " Instance Unavailable Please stand by!! \n";
          printf(" Your TAF type is %s \n",
                           (($fo_type==OCI_FO_NONE) ? "NONE"
                           :($fo_type==OCI_FO_SESSION) ? "SESSION"
                           :($fo_type==OCI_FO_SELECT) ? "SELECT"
                           : "UNKNOWN!"));
        }
        elsif ($fo_event == OCI_FO_ABORT){
           print " Failover aborted. Failover will not take place.\n";
        }
        elsif ($fo_event == OCI_FO_END){
           printf(" Failover ended ...Resuming your %s\n",(($fo_type==OCI_FO_NONE) ? "NONE"
                                                          :($fo_type==OCI_FO_SESSION) ? "SESSION"
                                                          :($fo_type==OCI_FO_SELECT) ? "SELECT"
                                                          : "UNKNOWN!"));
        }
        elsif ($fo_event == OCI_FO_REAUTH){
           print " Failed over user. Resuming services\n";
        }
        elsif ($fo_event == OCI_FO_ERROR){
           print " Failover error ...\n";
           sleep 5;                 # sleep before having another go
           return OCI_FO_RETRY;
        }
        else {
           printf(" Bad Failover Event: %d.\n",  $fo_event);

        }
        return 0;
      }

    The TAF types are as follows

      OCI_FO_SESSION indicates the user has requested only session failover.
      OCI_FO_SELECT indicates the user has requested select failover.
      OCI_FO_NONE indicates the user has not requested a failover type.
      OCI_FO_TXNAL indicates the user has requested a transaction failover.

    The TAF events are as follows

      OCI_FO_BEGIN indicates that failover has detected a lost connection and failover is starting.
      OCI_FO_END   indicates successful completion of failover.
      OCI_FO_ABORT indicates that failover was unsuccessful, and there is no option of retrying.
      OCI_FO_ERROR also indicates that failover was unsuccessful, but it gives the application the opportunity to handle the error and retry failover.
      OCI_FO_REAUTH indicates that you have multiple authentication handles and failover has occurred after the original authentication. It indicates that a user handle has been re-authenticated. To find out which, the application checks the OCI_ATTR_SESSION attribute of the service context handle (which is the first parameter).

   Connect Attributes

Changes
Revision history for DBD::Oracle

1.75_2    2014-11-19
  [ENHANCEMENT]

  Try and set -l when the build would have failed (H.Merijn Brand)

1.75_1    2014-11-17

  [DOCUMENTATION]

  Change mentions of READMEs in Makefile.PL to troubleshooting guides.
  (GH#17, reported by Ken Williams)

  [BUG FIXES]

  Fix GH#15 and GH#21 (the same problem). Previous change for
  RT91698 broke other things in output parameters.

1.74      2014-04-24
  - Promote to stable.

1.73_01   2014-04-23
  - Tweak fix for RT-88185. (GH#14, Martin J. Evans)

1.73_00   2014-04-23
  - Reverts current fix for RT-88185, as it causes breakage. (GH#14)

1.72      2014-04-14
  - promote 1.71_00 to stable.

1.71_00   2014-03-31
    - Recognizes __CYGWIN64__. (RT88709, reported by Witold Petriczek)
    - CHOOSE hint is deprecated. (RT91217, reported by Andy Bucksch,
        fix by Martin J Evans)
    - Set UTF8 flag per-connection. (RT88185, reported by Heinrich Mislik, patch by Martin
      J. Evans)
    - Add a CONTRIBUTING.mkd file. (GH#2)
    - Add SELinux trick. (RT#87003, patch submitted by Mike Doherty)

1.70   2014-02-12
  - promote 1.69_02 to stable.

1.69_02 2014-01-19

    [IMPROVEMENTS]

    - The DSN 'dbi:Oracle:sid=foo' is now an alias for 'dbi:Oracle:foo'.
        (RT#91775, Yanick Champoux, requested by David Wheeler)

    - Support for ORA_SYSBACKUP, ORA_SYSDG and ORA_SYSKM. (RT#91473,
        Kris Lemaire)

    [BUG FIXES]

   - OCI_THREADED setting had been accidentally removed, causing potential
    crashes when using threads. (RT#92229, Martin J. Evans, reported
    by Detlef Lütticke)

  - When using fetch*_hashref the values are decoded but
    not the keys so if you have unicode column names they were not
    returned correctly.  (RT#92134, Martin J. Evans, reported by
    Marcel Montes)


1.69_01 2014-01-14

  [BUG FIXES]

  - Fix RT91698. If you bound an output parameter to a scalar and
    repeatedly called execute the memory allocated in your bound
    scalar could increase each time. (Martin J. Evans)

1.68      2013-11-25
  - promote 1.67_00 to stable.

1.67_00   2013-11-05

  [BUG FIXES]
  - Fix RT88135. Add statistics_info support (patch by Steffen Goeldner)
  - Fix RT89491. Add RULE hint (patch by Steffen Goeldner)

  [DOCUMENTATION]
  - POD typos (RT#88285, RT#88284, Gregor Herrman).
  - Grooming of Hpux troubleshooting pod (GH#7, Martin J. Evans,
    Yanick Champoux)

1.66      2013-08-23
  - promote 1.65_00 to stable.

1.65_00   2013-07-29

  [BUG FIXES]

  - Fix RT85886. The TYPE passed to bind_col is supposed to be sticky
  and it wasn't. Attributes passed to bind_col could be lost later if
  bind_col is called again without attributes. Both of these occur
  when fetchall_arrayref is called with a slice (Martin J. Evans).

  [DOCUMENTATION]

  - Fix a bunch of typos. [GH#5, David Steinbrunner]

1.64      2013-05-22
  - promote 1.63_00 to stable.

1.63_00   2013-05-03

  [ENHANCEMENTS]
  - DBD-Oracle: Use of uninitialized value $user_only in uc [RT#84657]
    (Steffen Goeldner)

  [BUG FIXES]
  - Make 50cursor.t Oracle8-friendly. (RT#84660, patch by Steffen Goeldner)
  - Makefile.PL's use of ACL tweaked for Suse Enterprise 11 SP2
        (RT#84530, patch by Alfred Nathaniel)

  [DOCUMENTATION]
  - Bogus 227 directory no longer required for MacOS. (GH#1, patch
    by theory)

1.62      2013-04-30
  - promote 1.61_00 to official release

1.61_00   2013-04-15

  [BUG FIXES]
  - Adjust the privs needed for the DROP/CREATE table test. [GH#35]
    (Joe Crotty)

  - Fixed RT84170 - when using scrollable cursors and you've done a
    positioned fetch and then keep fetching until the end of the
    result-set calls to fetch never return undef and you keep getting
    the last row forever. Also added test case to the 51scroll.t test
    (Martin J. Evans).

1.60      2013-04-01
 - Move github repository to github.com/pythian/DBD-Oracle.

1.58      2013-03-05
  - promote 1.57_00 to official release

1.57_00   2013-02-07
  [BUG FIXES]

  - fix RT46628 - bind_param_inout ORA_RSET causes MSWin32 access
    violation and RT82663 - Errors if a returned SYS_REFCURSOR is not
    opened (Martin J. Evans)

  - Fix RT82663. If a procedure/function returns a SYS_REFCURSOR which
    was never opened DBD::Oracle magics a DBI statement handle into
    existence and attempts to describe it (which fails). This change
    examines a returned SYS_REFCURSOR and if it it is initialised but
    not executed does not create a DBI statement handle and returns
    undef instead. So now if you have a procedure/function which
    returns a SYS_REFCURSOR and never open it you'll get undef back
    instead of a useless statement handle.  Extended 50cursor.t test
    to check the above fix. (Martin J. Evans)

  [DOCUMENTATION]
  - Update Lion instructions for 64-bit Instant Client. (GH#37, patch by
    theory)

1.56      2013-01-08
  - fix t/26exe_array.t in the case of no db connection (RT82506,
    reported by Peter Rabbitson)

1.54      2013-01-03
 - promote 1.53_00 to official release

1.53_00      2012-12-18

  [BUG FIXES]
  - Fix RT69350 - 31lob.t was using $lob after destroying its parent $sth
        (Rob Davies)

  - Fix memory leak in execute_array (John Scoles, Pierre-Alain Blanc)

  - Fix RT80349 - The error message in execute_for_fetch when a row fails
    can contain the wrong error count. Thanks to Steffen Goeldner for
    RT and patch.

  - Fix RT80375 - no exception when execute_for_fetch fails and
    ArrayTupleStatus is not specified. Also the tuple count calculation
    resulted in an undefined warning. Thanks to Steffen Goeldner for
    RT and patch.

  - Fix RT80487. Skip XMLType tests if Oracle less than V9. Thanks to
    Steffen Goeldner for RT and patch.

  - Fix RT80486 for 31lob_extended.t. In old old Oracle8,
    SYS_REFCURSOR is not defined. Instead of CREATE/DROP PROCEDURE,
    use anonymous block. Thanks to Steffen Goeldner for RT and patch.


lib/DBD/Oracle/GetInfo.pm
package DBD::Oracle::GetInfo;
$DBD::Oracle::GetInfo::VERSION = '1.75_2';
our $AUTHORITY = 'cpan:PYTHIAN';
# ABSTRACT: Wrapper to get Oracle information

use strict;
use DBD::Oracle();

my $sql_driver = 'Oracle';
my $sql_ver_fmt = '%02d.%02d.%04d';   # ODBC version string: ##.##.#####
my ($a,$b,$c) = (0,0,0);
my $ver = $DBD::Oracle::VERSION;
my @parts = split /\./, $ver;
$a = $parts[0];
($b,$c) = split /_/, $parts[1];
$c = 0 if !$c;

my $sql_driver_ver = sprintf $sql_ver_fmt, $a, $b, $c;

sub sql_dbms_version {
    my $dbh = shift;
    local $^W; # for ora_server_version having too few parts
    return sprintf $sql_ver_fmt, @{DBD::Oracle::db::ora_server_version($dbh)};
}

my @Keywords = qw(
ACCESS
AUDIT
BFILE
BLOB
CLOB
CLUSTER
COMMENT
COMPRESS
EXCLUSIVE
FILE
IDENTIFIED
INCREMENT
INITIAL
LOCK
LONG
MAXEXTENTS
MINUS
MODE
MODIFY
NCLOB
NOAUDIT
NOCOMPRESS
NOWAIT
NUMBER
OFFLINE
ONLINE
PCTFREE
RAW
RENAME
RESOURCE
ROW
ROWID
ROWLABEL
ROWNUM
SHARE
START
SUCCESSFUL
SYNONYM
SYSDATE
TRIGGER
UID
VALIDATE
VARCHAR
);


sub sql_keywords {

    return join ',', @Keywords;

}



sub sql_data_source_name {
    my $dbh = shift;
    return "dbi:$sql_driver:" . $dbh->{Name};
}

sub sql_user_name {
    my $dbh = shift;
    # CURRENT_USER is a non-standard attribute, probably undef
    # Username is a standard DBI attribute
    return $dbh->{CURRENT_USER} || $dbh->{Username};
}


our %info = (
      0 => 0,                             # SQL_MAX_DRIVER_CONNECTIONS
      1 => 0,                             # SQL_MAX_CONCURRENT_ACTIVITIES
      2 => \&sql_data_source_name,        # SQL_DATA_SOURCE_NAME
      3 => 147209344,                     # SQL_DRIVER_HDBC
      4 => 147212776,                     # SQL_DRIVER_HENV
#     5 => undef,                         # SQL_DRIVER_HSTMT
      6 => $INC{'DBD/Oracle.pm'},         # SQL_DRIVER_NAME
      7 => $sql_driver_ver,               # SQL_DRIVER_VER
      8 => 191,                           # SQL_FETCH_DIRECTION
      9 => 1,                             # SQL_ODBC_API_CONFORMANCE
     10 => '03.52',                       # SQL_ODBC_VER
     11 => 'Y',                           # SQL_ROW_UPDATES
     12 => 0,                             # SQL_ODBC_SAG_CLI_CONFORMANCE
     13 => sub {"$_[0]->{Name}"},         # SQL_SERVER_NAME
     14 => '\\',                          # SQL_SEARCH_PATTERN_ESCAPE
     15 => 1,                             # SQL_ODBC_SQL_CONFORMANCE
     16 => 'DEVEL',                       # SQL_DATABASE_NAME
     17 => 'Oracle',                      # SQL_DBMS_NAME
     18 => \&sql_dbms_version,            # SQL_DBMS_VERSION
     19 => 'Y',                           # SQL_ACCESSIBLE_TABLES
     20 => 'Y',                           # SQL_ACCESSIBLE_PROCEDURES
     21 => 'Y',                           # SQL_PROCEDURES
     22 => 1,                             # SQL_CONCAT_NULL_BEHAVIOR
     23 => 2,                             # SQL_CURSOR_COMMIT_BEHAVIOR
     24 => 2,                             # SQL_CURSOR_ROLLBACK_BEHAVIOR
     25 => 'N',                           # SQL_DATA_SOURCE_READ_ONLY
     26 => 8,                             # SQL_DEFAULT_TRANSACTION_ISOLATION
     27 => 'Y',                           # SQL_EXPRESSIONS_IN_ORDERBY
     28 => 1,                             # SQL_IDENTIFIER_CASE
     29 => '"',                           # SQL_IDENTIFIER_QUOTE_CHAR
     30 => 30,                            # SQL_MAXIMUM_COLUMN_NAME_LENGTH
     31 => 30,                            # SQL_MAXIMUM_CURSOR_NAME_LENGTH
     32 => 30,                            # SQL_MAXIMUM_SCHEMA_NAME_LENGTH
     33 => 92,                            # SQL_MAX_PROCEDURE_NAME_LEN
     34 => 0,                             # SQL_MAXIMUM_CATALOG_NAME_LENGTH
     35 => 30,                            # SQL_MAXIMUM_TABLE_NAME_LENGTH
     36 => 'Y',                           # SQL_MULT_RESULT_SETS
     37 => 'Y',                           # SQL_MULTIPLE_ACTIVE_TXN
     38 => 'Y',                           # SQL_OUTER_JOINS
     39 => 'Owner',                       # SQL_SCHEMA_TERM
     40 => 'Procedure',                   # SQL_PROCEDURE_TERM
     41 => '@',                           # SQL_QUALIFIER_NAME_SEPARATOR
     42 => 'Database Link',               # SQL_QUALIFIER_TERM
     43 => 7,                             # SQL_SCROLL_CONCURRENCY
     44 => 19,                            # SQL_SCROLL_OPTIONS
     45 => 'Table',                       # SQL_TABLE_TERM
     46 => 3,                             # SQL_TRANSACTION_CAPABLE
     47 => \&sql_user_name,               # SQL_USER_NAME
     48 => 1,                             # SQL_CONVERT_FUNCTIONS
     49 => 16646015,                      # SQL_NUMERIC_FUNCTIONS
     50 => 8355839,                       # SQL_STRING_FUNCTIONS
     51 => 7,                             # SQL_SYSTEM_FUNCTIONS
     52 => 1023999,                       # SQL_TIMEDATE_FUNCTIONS
     53 => 10518015,                      # SQL_CONVERT_BIGINT
     54 => 10775839,                      # SQL_CONVERT_BINARY
     55 => 10518015,                      # SQL_CONVERT_BIT
     56 => 15106047,                      # SQL_CONVERT_CHAR
     57 => 164097,                        # SQL_CONVERT_DATE
     58 => 10518015,                      # SQL_CONVERT_DECIMAL
     59 => 10514943,                      # SQL_CONVERT_DOUBLE
     60 => 10514943,                      # SQL_CONVERT_FLOAT
     61 => 10518015,                      # SQL_CONVERT_INTEGER
     62 => 14680833,                      # SQL_CONVERT_LONGVARCHAR
     63 => 10518015,                      # SQL_CONVERT_NUMERIC
     64 => 10514943,                      # SQL_CONVERT_REAL
     65 => 10518015,                      # SQL_CONVERT_SMALLINT
     66 => 0,                             # SQL_CONVERT_TIME
     67 => 10718465,                      # SQL_CONVERT_TIMESTAMP
     68 => 10518015,                      # SQL_CONVERT_TINYINT
     69 => 10775839,                      # SQL_CONVERT_VARBINARY
     70 => 15204351,                      # SQL_CONVERT_VARCHAR
     71 => 265216,                        # SQL_CONVERT_LONGVARBINARY
     72 => 10,                            # SQL_TRANSACTION_ISOLATION_OPTION
     73 => 'N',                           # SQL_ODBC_SQL_OPT_IEF
     74 => 2,                             # SQL_CORRELATION_NAME
     75 => 1,                             # SQL_NON_NULLABLE_COLUMNS
#    76 => undef,                         # SQL_DRIVER_HLIB
     77 => '03.52',                       # SQL_DRIVER_ODBC_VER
     78 => 1,                             # SQL_LOCK_TYPES
     79 => 1,                             # SQL_POS_OPERATIONS
     80 => 7,                             # SQL_POSITIONED_STATEMENTS
     81 => 15,                            # SQL_GETDATA_EXTENSIONS
     82 => 88,                            # SQL_BOOKMARK_PERSISTENCE
     83 => 0,                             # SQL_STATIC_SENSITIVITY
     84 => 0,                             # SQL_FILE_USAGE
     85 => 1,                             # SQL_NULL_COLLATION
     86 => 1029739,                       # SQL_ALTER_TABLE
     87 => 'Y',                           # SQL_COLUMN_ALIAS
     88 => 2,                             # SQL_GROUP_BY
     89 => \&sql_keywords,                # SQL_KEYWORDS
     90 => 'N',                           # SQL_ORDER_BY_COLUMNS_IN_SELECT
     91 => 31,                            # SQL_SCHEMA_USAGE
     92 => 3,                             # SQL_QUALIFIER_USAGE
     93 => 3,                             # SQL_QUOTED_IDENTIFIER_CASE
     94 => '$#',                          # SQL_SPECIAL_CHARACTERS
     95 => 31,                            # SQL_SUBQUERIES
     96 => 3,                             # SQL_UNION_STATEMENT
     97 => 0,                             # SQL_MAXIMUM_COLUMNS_IN_GROUP_BY
     98 => 0,                             # SQL_MAXIMUM_COLUMNS_IN_INDEX
     99 => 0,                             # SQL_MAXIMUM_COLUMNS_IN_ORDER_BY
    100 => 1000,                          # SQL_MAXIMUM_COLUMNS_IN_SELECT
    101 => 1000,                          # SQL_MAXIMUM_COLUMNS_IN_TABLE
    102 => 0,                             # SQL_MAXIMUM_INDEX_SIZE
    103 => 'N',                           # SQL_MAX_ROW_SIZE_INCLUDES_LONG
    104 => 0,                             # SQL_MAXIMUM_ROW_SIZE
    105 => 0,                             # SQL_MAXIMUM_STATEMENT_LENGTH
    106 => 0,                             # SQL_MAXIMUM_TABLES_IN_SELECT
    107 => 30,                            # SQL_MAXIMUM_USER_NAME_LENGTH
    108 => 0,                             # SQL_MAX_CHAR_LITERAL_LEN
    109 => 0,                             # SQL_TIMEDATE_ADD_INTERVALS
    110 => 0,                             # SQL_TIMEDATE_DIFF_INTERVALS
    111 => 'N',                           # SQL_NEED_LONG_DATA_LEN
    112 => 0,                             # SQL_MAX_BINARY_LITERAL_LEN
    113 => 'Y',                           # SQL_LIKE_ESCAPE_CLAUSE
    114 => 2,                             # SQL_QUALIFIER_LOCATION
    115 => 127,                           # SQL_OUTER_JOIN_CAPABILITIES
    116 => 0,                             # SQL_ACTIVE_ENVIRONMENTS
    117 => 0,                             # SQL_ALTER_DOMAIN
    118 => 1,                             # SQL_SQL_CONFORMANCE
    119 => 0,                             # SQL_DATETIME_LITERALS
    120 => 0,                             # SQL_BATCH_ROW_COUNT
    121 => 0,                             # SQL_BATCH_SUPPORT
    122 => 15106047,                      # SQL_CONVERT_WCHAR
    123 => 0,                             # SQL_CONVERT_INTERVAL_DAY_TIME
    124 => 0,                             # SQL_CONVERT_INTERVAL_YEAR_MONTH
    125 => 14680833,                      # SQL_CONVERT_WLONGVARCHAR
    126 => 15106047,                      # SQL_CONVERT_WVARCHAR
    127 => 0,                             # SQL_CREATE_ASSERTION
    128 => 0,                             # SQL_CREATE_CHARACTER_SET
    129 => 0,                             # SQL_CREATE_COLLATION
    130 => 0,                             # SQL_CREATE_DOMAIN
    131 => 3,                             # SQL_CREATE_SCHEMA
    132 => 14305,                         # SQL_CREATE_TABLE
    133 => 0,                             # SQL_CREATE_TRANSLATION
    134 => 3,                             # SQL_CREATE_VIEW
#   135 => undef,                         # SQL_DRIVER_HDESC
    136 => 0,                             # SQL_DROP_ASSERTION
    137 => 0,                             # SQL_DROP_CHARACTER_SET
    138 => 0,                             # SQL_DROP_COLLATION
    139 => 0,                             # SQL_DROP_DOMAIN
    140 => 0,                             # SQL_DROP_SCHEMA
    141 => 1,                             # SQL_DROP_TABLE
    142 => 0,                             # SQL_DROP_TRANSLATION
    143 => 1,                             # SQL_DROP_VIEW
    144 => 0,                             # SQL_DYNAMIC_CURSOR_ATTRIBUTES1
    145 => 0,                             # SQL_DYNAMIC_CURSOR_ATTRIBUTES2
    146 => 57345,                         # SQL_FORWARD_ONLY_CURSOR_ATTRIBUTES1
    147 => 2183,                          # SQL_FORWARD_ONLY_CURSOR_ATTRIBUTES2
    148 => 3,                             # SQL_INDEX_KEYWORDS
    149 => 65568,                         # SQL_INFO_SCHEMA_VIEWS
    150 => 0,                             # SQL_KEYSET_CURSOR_ATTRIBUTES1
    151 => 0,                             # SQL_KEYSET_CURSOR_ATTRIBUTES2
    152 => 3,                             # SQL_ODBC_INTERFACE_CONFORMANCE
    153 => 2,                             # SQL_PARAM_ARRAY_ROW_COUNTS
    154 => 3,                             # SQL_PARAM_ARRAY_SELECTS
    155 => 0,                             # SQL_SQL92_DATETIME_FUNCTIONS
    156 => 0,                             # SQL_SQL92_FOREIGN_KEY_DELETE_RULE
    157 => 0,                             # SQL_SQL92_FOREIGN_KEY_UPDATE_RULE
    158 => 16,                            # SQL_SQL92_GRANT
    159 => 0,                             # SQL_SQL92_NUMERIC_VALUE_FUNCTIONS
    160 => 7687,                          # SQL_SQL92_PREDICATES
    161 => 0,                             # SQL_SQL92_RELATIONAL_JOIN_OPERATORS
    162 => 0,                             # SQL_SQL92_REVOKE
    163 => 0,                             # SQL_SQL92_ROW_VALUE_CONSTRUCTOR
    164 => 0,                             # SQL_SQL92_STRING_FUNCTIONS
    165 => 1,                             # SQL_SQL92_VALUE_EXPRESSIONS
    166 => 3,                             # SQL_STANDARD_CLI_CONFORMANCE
    167 => 57935,                         # SQL_STATIC_CURSOR_ATTRIBUTES1
    168 => 4231,                          # SQL_STATIC_CURSOR_ATTRIBUTES2
    169 => 64,                            # SQL_AGGREGATE_FUNCTIONS
    170 => 3,                             # SQL_DDL_INDEX
    171 => '03.52.0002.0002',             # SQL_DM_VER
    172 => 7,                             # SQL_INSERT_STATEMENT
    173 => 0,                             # SQL_CONVERT_GUID
  10000 => 1995,                          # SQL_XOPEN_CLI_YEAR
  10001 => 1,                             # SQL_CURSOR_SENSITIVITY
  10002 => 'Y',                           # SQL_DESCRIBE_PARAMETER
  10003 => 'N',                           # SQL_CATALOG_NAME
  10004 => '',                            # SQL_COLLATING_SEQUENCE
  10005 => 30,                            # SQL_MAXIMUM_IDENTIFIER_LENGTH
  10021 => 2,                             # SQL_ASYNC_MODE
  10022 => 0,                             # SQL_MAX_ASYNC_CONCURRENT_STATEMENTS
# 20000 => undef,                         # SQL_MAXIMUM_STMT_OCTETS
# 20001 => undef,                         # SQL_MAXIMUM_STMT_OCTETS_DATA
# 20002 => undef,                         # SQL_MAXIMUM_STMT_OCTETS_SCHEMA
);

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

DBD::Oracle::GetInfo - Wrapper to get Oracle information

=head1 VERSION

version 1.75_2

=head1 AUTHORS

=over 4

=item *

Tim Bunce <timb@cpan.org>

=item *

John Scoles <byterock@cpan.org>

=item *

Yanick Champoux <yanick@cpan.org>

=item *

Martin J. Evans <mjevans@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 1994 by Tim Bunce.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
